### PR TITLE
Zcash-style serialization for BLS12-381

### PR DESCRIPTION
**Pull request by mmagician**
## Description

Adapted from https://github.com/crate-crypto/small-powers-of-tau/blob/master/src/interop_point_encoding.rs to take advantage of [flexible serialization for curves](https://github.com/arkworks-rs/algebra/pull/467).

Question: should we maybe have this mode of serialization feature-gated, e.g.:
```
...
#[cfg(feature = "zcash_serialization")]
fn serialized_size(_compress: ark_serialize::Compress) -> usize {
    G2_SERIALISED_SIZE
}
...
```

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: bls12_381/src/curves/g1.rs [hunk 1]
```diff
@@ -62,6 +63,68 @@ impl SWCurveConfig for Parameters {
...
+        G1Affine::get_point_from_x_unchecked(x, flags.is_lexographically_largest)
+            .ok_or(SerializationError::InvalidData)
+    }
+
+    fn serialize_with_mode<W: ark_serialize::Write>(
+        item: &Affine<Self>,
+        mut writer: W,
+        _compress: ark_serialize::Compress,
+    ) -> Result<(), SerializationError> {
+        let encoding = EncodingFlags {
```
Comment (Pratyush) [inline bls12_381/src/curves/g1.rs:111]: Same here; we should support both compressed and uncompressed serialization.

File: bls12_381/src/curves/g1.rs [hunk 2]
```diff
@@ -62,6 +63,68 @@ impl SWCurveConfig for Parameters {
         let endomorphism_p = endomorphism(p);
         minus_x_squared_times_p.eq(&endomorphism_p)
     }
+
+    // modes are ignored, as the flags are already encoded in the data
+    fn deserialize_with_mode<R: ark_serialize::Read>(
+        mut reader: R,
+        _compress: ark_serialize::Compress,
+        _validate: ark_serialize::Validate,
```
Comment (Pratyush) [inline bls12_381/src/curves/g1.rs:71]: We should use `compress` to check that it matches the encoded flag, and `validate` to conditionally perform a subgroup check.

File: bls12_381/src/curves/g2.rs [hunk 3]
```diff
@@ -69,6 +70,78 @@ impl SWCurveConfig for Parameters {
 
         x_times_point.eq(&p_times_point)
     }
+
+    fn deserialize_with_mode<R: ark_serialize::Read>(
+        mut reader: R,
+        _compress: ark_serialize::Compress,
+        _validate: ark_serialize::Validate,
```
Comment (Pratyush) [inline bls12_381/src/curves/g2.rs:77]: We should use `compress` to check that it matches the encoded flag, and `validate` to conditionally perform a subgroup check.

File: bls12_381/src/curves/g1.rs (hunk 2 above)
Comment (mmagician) [reply bls12_381/src/curves/g1.rs]: Done

File: bls12_381/src/curves/g2.rs [hunk 4]
```diff
@@ -69,6 +70,78 @@ impl SWCurveConfig for Parameters {
...
+        G2Affine::get_point_from_x_unchecked(x, flags.is_lexographically_largest)
+            .ok_or(SerializationError::InvalidData)
+    }
+
+    fn serialize_with_mode<W: ark_serialize::Write>(
+        item: &Affine<Self>,
+        mut writer: W,
+        _compress: ark_serialize::Compress,
+    ) -> Result<(), SerializationError> {
+        let encoding = EncodingFlags {
```
[the remark this answers is not in the document]
Comment (mmagician) [reply bls12_381/src/curves/g2.rs:98]: Done, both compression modes are now supported.

File: bls12_381/src/curves/g2.rs (hunk 3 above)
Comment (mmagician) [reply bls12_381/src/curves/g2.rs]: Done

**Comment by mmagician** [comment]
@Pratyush Thanks for the review, these are of course __valid points__ 😉 
As you've seen, I've opened another [fix](https://github.com/arkworks-rs/algebra/pull/495) for `get_point_from_x_unchecked` - this PR depends on it for the tests to pass.

File: bls12_381/src/curves/g1.rs [hunk 5]
```diff
@@ -62,6 +63,68 @@ impl SWCurveConfig for Parameters {
         let endomorphism_p = endomorphism(p);
         minus_x_squared_times_p.eq(&endomorphism_p)
     }
+
+    // modes are ignored, as the flags are already encoded in the data
```
Comment (Pratyush) [inline bls12_381/src/curves/g1.rs:67]:
```suggestion
```

File: bls12_381/src/curves/g1.rs [hunk 6]
```diff
@@ -62,6 +63,68 @@ impl SWCurveConfig for Parameters {
...
+        compress: ark_serialize::Compress,
+    ) -> Result<(), SerializationError> {
+        let encoding = EncodingFlags {
+            is_compressed: compress == ark_serialize::Compress::Yes,
+            is_infinity: item.is_zero(),
+            is_lexographically_largest: item.y > -item.y,
+        };
+        let mut p = *item;
+        if encoding.is_infinity {
+            p = G1Affine::default();
```
Comment (Pratyush) [inline bls12_381/src/curves/g1.rs:99]:
```suggestion
            p = G1Affine::zero();
```

File: bls12_381/src/curves/mod.rs [hunk 7]
```diff
@@ -28,3 +32,84 @@ impl Bls12Parameters for Parameters {
     type G1Parameters = self::g1::Parameters;
     type G2Parameters = self::g2::Parameters;
 }
+
```
Comment (Pratyush) [inline bls12_381/src/curves/mod.rs:35]: Maybe let's put these serialization related structs in a separate file?

File: bls12_381/src/curves/g1.rs [hunk 8]
```diff
@@ -83,3 +146,99 @@ pub fn endomorphism(p: &Affine<Parameters>) -> Affine<Parameters> {
...
+#[cfg(test)]
+mod tests {
+    use ark_ec::AffineRepr;
+    use ark_serialize::{CanonicalDeserialize, CanonicalSerialize, Compress};
+
+    use crate::G1Affine;
+    extern crate alloc;
+    use alloc::vec;
+
+    #[test]
```
Comment (Pratyush) [inline bls12_381/src/curves/g1.rs:220]: Are there any standard test vectors that we could lint against here?

**Comment by Pratyush** [comment]
btw I think there's no need to feature-gate the new compression. We can just leave a note in the CHANGELOG about this being a breaking change.

File: bls12_381/src/curves/mod.rs (hunk 7 above)
Comment (mmagician) [reply bls12_381/src/curves/mod.rs]: 👍🏼 done in cd97b6a2a6120f5d978015b36d83d7d02e62ab88 and 205096c0eec479505e88d37fdeda7e9f34d16979

File: bls12_381/src/curves/g1.rs (hunk 8 above)
Comment (mmagician) [reply bls12_381/src/curves/g1.rs]: There are indeed, I imported them directly from https://github.com/zkcrypto/bls12_381 in c7a53c6.
They also encompass the previous hardcoded tests for zero & generator: they're the first two entries of the test vectors, so I've removed the hardcoded ones grom `g1.rs` and `g2.rs`.

File: bls12_381/src/curves/g1.rs [hunk 9]
```diff
@@ -7,7 +7,13 @@ use ark_ec::{
     AffineRepr, Group,
 };
 use ark_ff::{Field, MontFp, PrimeField, Zero};
+use ark_serialize::{Compress, SerializationError};
 use ark_std::{ops::Neg, One};
+
+use crate::util::{
+    read_g1_compressed, read_g1_uncompressed, serialise_fq, EncodingFlags, G1_SERIALISED_SIZE,
```
Comment (daira) [inline bls12_381/src/curves/g1.rs:14]: The spelling of "seriali[sz]e" is inconsistent. (One of the reasons that, as a Brit, I have always still used US spelling in code.)

Given that this is a breaking change anyway, maybe it is a chance to make this consistent?

File: bls12_381/src/curves/g1.rs [hunk 10]
```diff
@@ -70,6 +76,66 @@ impl SWCurveConfig for Parameters {
...
+        };
+
+        Ok(())
+    }
+
+    fn serialized_size(compress: Compress) -> usize {
+        if compress == Compress::Yes {
+            G1_SERIALISED_SIZE
+        } else {
+            G1_SERIALISED_SIZE * 2
```
Comment (daira) [inline bls12_381/src/curves/g1.rs:136]: Note that Zcash always uses compressed point encodings, so the uncompressed version should not be referred to as Zcash-style encoding in documentation.

File: bls12_381/src/curves/g1.rs (hunk 9 above)
Comment (mmagician) [reply bls12_381/src/curves/g1.rs]: Good call, [done](https://github.com/arkworks-rs/curves/pull/129/commits/d8b33388f7a55627c42e39096e72ed164718c6dd).

File: bls12_381/src/curves/g1.rs (hunk 10 above)
Comment (mmagician) [reply bls12_381/src/curves/g1.rs]: @daira Noted, thanks. So far it's not mentioned in the docs, but will go into the [CHANGELOG](https://github.com/arkworks-rs/curves/pull/131/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR11). I've just reworded it there, I hope it's fine like that?

File: bls12_381/src/curves/util.rs [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+pub(crate) fn read_g1_uncompressed<R: ark_serialize::Read>(
+    mut reader: R,
+) -> Result<Affine<G1Parameters>, ark_serialize::SerializationError> {
+    let mut bytes = [0u8; 2 * G1_SERIALIZED_SIZE];
+    reader
+        .read_exact(&mut bytes)
+        .ok()
+        .ok_or(SerializationError::InvalidData)?;
```
Comment (Pratyush) [inline bls12_381/src/curves/util.rs:129]:
```suggestion
        .map_err(|_| SerializationError::InvalidData)?;
```

File: bls12_381/src/curves/util.rs [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+pub(crate) fn read_g2_compressed<R: ark_serialize::Read>(
+    mut reader: R,
+) -> Result<Affine<G2Parameters>, ark_serialize::SerializationError> {
+    let mut bytes = [0u8; G2_SERIALIZED_SIZE];
+    reader
+        .read_exact(&mut bytes)
+        .ok()
+        .ok_or(SerializationError::InvalidData)?;
```
Comment (Pratyush) [inline bls12_381/src/curves/util.rs:160]:
```suggestion
        .map_err(|_| SerializationError::InvalidData)?;
```

File: bls12_381/src/curves/util.rs [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+pub(crate) fn read_g2_uncompressed<R: ark_serialize::Read>(
+    mut reader: R,
+) -> Result<Affine<G2Parameters>, ark_serialize::SerializationError> {
+    let mut bytes = [0u8; 2 * G2_SERIALIZED_SIZE];
+    reader
+        .read_exact(&mut bytes)
+        .ok()
+        .ok_or(SerializationError::InvalidData)?;
```
Comment (Pratyush) [inline bls12_381/src/curves/util.rs:193]:
```suggestion
        .map_err(|_| SerializationError::InvalidData)?;
```

**Review by Pratyush — APPROVED** [review]
